# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/clients/manager/ClientManager.cpp
# Commit: 9e8267a443e5
# Full Hash: 9e8267a443e521f496486a183298fc23ebe8ee15
# Author: Luca Greco <lgreco@mozilla.com>
# Date: 2021-09-01 03:48:05
# Regressor Bug: 1727405
# File Overlap Count: 1
# Description:
#   Bug 1727405 - Add a diagnostic assert to check that ClientManager::GetOrCreateForCurrentThread is not returning a stale ClientManager instance that is already shutdown. r=asuth
#   
#   While investigating an unexpected test failure triggered by a mochitest that was testing the identity WebExtensions API
#   (D121683 from Bug 1723852) I did notice that the actual underlying issue was triggered by a leak
#   (in particular the Extension API class in the initial draft of that D121683 patch was missing a RefPtr
# ==============================================================================

diff -r e8a630d41a77 -r 9e8267a443e5 dom/clients/manager/ClientManager.cpp
--- a/dom/clients/manager/ClientManager.cpp	Tue Aug 31 18:40:20 2021 +0300
+++ b/dom/clients/manager/ClientManager.cpp	Tue Aug 31 14:18:35 2021 +0000
@@ -217,6 +217,12 @@
   }
 
   MOZ_DIAGNOSTIC_ASSERT(cm);
+  // Check that the ClientManager instance associated to the current thread has
+  // not been kept alive when it was expected to have been already deallocated
+  // (e.g. due to a leak ClientManager's mShutdown can have ben set to true from
+  // its RevokeActor method but never fully deallocated and unset from the
+  // thread locals).
+  MOZ_DIAGNOSTIC_ASSERT(!cm->IsShutdown());
   return cm.forget();
 }
 
